Initial classes

mecha opened the pull request:
Initial classes for a basic event-driven state machine implementation.

The machine uses an event manager to trigger events on transition. A specialized `TransitionEvent` instance is given to the event manager. Handlers can use the event to abort transitions ~and also specify the new state of the machine.~ The transition key is used as the new state, unless the transition is aborted.

If an exception is thrown, the transition will still happen. Event handlers will need to catch exceptions and explicitly abort the transition.

mecha committed:
Added PHPStorm code style config

mecha committed:
Installed dependencies

- State machine interfaces
- State machine abstract implementations
- I18n helper traits
- Exception helper traits
- Transition event interfaces
- PSR-14 Event Manager interfaces (dev)
- Stringable interfces (dev)

mecha committed:
Added PHPUnit config for PHPStorm

mecha committed:
Added `AbstractStateMachineException`

mecha committed:
Added test for `StateMachineException`

mecha committed:
Added `StateMachineException`

mecha committed:
Added test for `CouldNotTransitionException`

mecha committed:
Added `CouldNotTransitionException`

mecha committed:
Added test for `TransitionEvent`

mecha committed:
Added `TransitionEvent`

mecha committed:
Added test for `EventStateMachine`

mecha committed:
Added `EventStateMachine`

mecha committed:
Added test for `AbstractEventStateMachine`

mecha committed:
Added `AbstractEventStateMachine`

mecha committed:
`EventStateMachine` extends `AbstractEventStateMachine`

mecha committed:
Updated test for `EventStateMachine`

mecha committed:
Removed unusued use statement

mecha committed:
Default event name pattern does not contain transition

mecha committed:
Updated PHP doc block for `EventStateMachine`

mecha committed:
Updated dependencies

mecha committed:
Updated dependencies

mecha committed:
Merge branch 'task/initial-classes' of https://github.com/rebelcode/event-state-machine into task/initial-classes

XedinUnknown commented on src/AbstractEventStateMachine.php:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        if (isset($exception) && $exception instanceof Exception) {
+            throw $exception;
```
Nothing is apparently thrown if event handlers did not throw. However, the interface, AFAIK, [requires](https://github.com/Dhii/machine-state-machine-interface/blob/develop/src/StateMachineInterface.php#L31) one of two things happen: the transition happens, or [`CouldNotTransitionExceptionInterface`] is thrown. Therefore, the exception must be thrown when the transition is aborted, regardless of whether the event mechanism threw.

[`CouldNotTransitionExceptionInterface`]: https://github.com/Dhii/machine-state-machine-interface/blob/develop/src/Exception/CouldNotTransitionExceptionInterface.php

XedinUnknown commented on src/EventStateMachine.php:
```diff
@@ -0,0 +1,352 @@
+ *
+ * This implementation does not use an internal state graph. Instead, transition given to be applied will be
+ * used as the new state. As such, it is perfectly valid to transition to the same state, thus making this state
+ * machine implementation non-deterministic.
```
Looks deterministic to me.

mecha replied:
Forgot to remove after changing the mechanics. Thanks for pointing out.

mecha committed:
Updated description for `EventStateMachine`

The class doc incorrectly stated that the machine is non-deterministic.

mecha committed:
Added abstract factory method for creating ...

... `StateMachineExceptionInterface` instances.

mecha committed:
Added missing docs in `AbstractEventStateMachine`

mecha committed:
Fixed exceptions in `_trigger()`

Now correctly throws exception for all scenarios:

If the transition was aborted, a `CouldNotTransitionExceptionInterface` is *always* thrown. If an exception is thrown by the event handlers , it will be attached as an inner exception.
If the transition was not aborted and an exception was thrown by an event handler, the transition will be committed and a `StateMachineExceptionInterface` is thrown with the event handler's exception attached as an inner exception.

XedinUnknown commented on src/AbstractEventStateMachine.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return CouldNotTransitionExceptionInterface The created exception.
+     */
+    abstract protected function _createStateMachineException(
```
Not used in this class. Why is it here?

mecha replied:
Used in [`AbstractEventStateMachine#46`](https://github.com/RebelCode/event-state-machine/blob/864cbf673fe18f93ef97b775b8894f562c95c749/src/AbstractEventStateMachine.php#L46)

XedinUnknown replied:
My bad.
